File: integrations/old-source-preservation/ivault-eye-tracking/snapshot/src/features/remote/remotePolicy.ts
import type {
  RemoteCommand,
  RemoteCommandType,
  RemoteInputSource,
  RemotePermissionResult,
  RemoteRiskLevel,
  RemoteUserPolicyContext,
} from "./types";

export const DEFAULT_REMOTE_POLICY_CONTEXT: RemoteUserPolicyContext = {
  ageGroup: "adult",
  kycVerified: true,
  trustTier: 2,
  walletLocked: false,
  fraudHoldActive: false,
  campaignBudgetAvailable: true,
  canPublishCampaign: true,
  canWithdraw: true,
  canPay: true,
  canTip: true,
};

const HIGH_RISK_COMMAND_TYPES = new Set<RemoteCommandType>([
  "WITHDRAW",
  "PAY",
  "TIP",
  "PUBLISH_CAMPAIGN",
  "CONVERT_COINS",
]);

const GAZE_BLOCKED_COMMAND_TYPES = new Set<RemoteCommandType>([
  "WITHDRAW",
  "PAY",
  "TIP",
  "PUBLISH_CAMPAIGN",
  "CONNECT_PLATFORM",
  "DISCONNECT_PLATFORM",
  "CONVERT_COINS",
]);

const MONEY_COMMAND_TYPES = new Set<RemoteCommandType>([
  "WITHDRAW",
  "PAY",
  "TIP",
  "CONVERT_COINS",
]);

const CONNECTOR_MUTATION_TYPES = new Set<RemoteCommandType>([
  "CONNECT_PLATFORM",
  "DISCONNECT_PLATFORM",
]);

export function evaluateRemotePermission(
  command: RemoteCommand,
  policy: RemoteUserPolicyContext,
  remoteLocked: boolean
): RemotePermissionResult {
  if (remoteLocked && command.type !== "UNLOCK_REMOTE") {
    return { allowed: false, reason: "Remote is locked.", requiresConfirmation: false };
  }

  if (policy.ageGroup === "minor" && command.requiresAdult) {
    return {
      allowed: false,
      reason: "This action is not available for this account.",
      requiresConfirmation: false,
    };
  }

  if (command.requiresKyc && !policy.kycVerified) {
    return {
      allowed: false,
      reason: "Identity verification required.",
      requiresConfirmation: false,
    };
  }

  if (
    command.trustTierRequired != null &&
    policy.trustTier < command.trustTierRequired
  ) {
    return { allowed: false, reason: "Trust tier too low.", requiresConfirmation: false };
  }

  if (policy.fraudHoldActive) {
    return {
      allowed: false,
      reason: "Action paused by safety review.",
      requiresConfirmation: false,
    };
  }

  if (command.type === "PUBLISH_CAMPAIGN" && !policy.canPublishCampaign) {
    return {
      allowed: false,
      reason: "Campaign publishing not allowed.",
      requiresConfirmation: false,
    };
  }

  if (!policy.campaignBudgetAvailable && command.type === "PUBLISH_CAMPAIGN") {
    return {
      allowed: false,
      reason: "Campaign budget not ready.",
      requiresConfirmation: false,
    };
  }

  if (
    (command.type === "PAY" && !policy.canPay) ||
    (command.type === "WITHDRAW" && !policy.canWithdraw) ||
    (command.type === "TIP" && !policy.canTip)
  ) {
    return { allowed: false, reason: "Wallet action not allowed.", requiresConfirmation: false };
  }

  if (
    (command.type === "PAY" ||
      command.type === "WITHDRAW" ||
      command.type === "TIP") &&
    policy.walletLocked
  ) {
    return { allowed: false, reason: "Wallet is locked.", requiresConfirmation: false };
  }

  if (command.inputSource === "gaze" && GAZE_BLOCKED_COMMAND_TYPES.has(command.type)) {
    return {
      allowed: false,
      reason: "This action cannot run from gaze input.",
      requiresConfirmation: false,
    };
  }

  const requiresConfirmation =
    command.riskLevel === "HIGH" ||
    command.riskLevel === "MEDIUM" ||
    command.requiresConfirmation === true;

  return { allowed: true, requiresConfirmation };
}

export function getRemoteRiskLevelForInput(
  inputSource: RemoteInputSource,
  commandType: RemoteCommandType
): RemoteRiskLevel {
  if (inputSource === "gaze" && GAZE_BLOCKED_COMMAND_TYPES.has(commandType)) {
    return "BLOCKED";
  }
  if (HIGH_RISK_COMMAND_TYPES.has(commandType)) {
    return "HIGH";
  }
  return "LOW";
}

export function buildConfirmationCopy(command: RemoteCommand): string {
  return `Confirm "${command.label}" (${command.type})? Risk: ${command.riskLevel}.`;
}

export function isMoneyCommand(type: RemoteCommandType): boolean {
  return MONEY_COMMAND_TYPES.has(type);
}

export function isCampaignSpendCommand(type: RemoteCommandType): boolean {
  return type === "PUBLISH_CAMPAIGN";
}

export function isConnectorMutationCommand(type: RemoteCommandType): boolean {
  return CONNECTOR_MUTATION_TYPES.has(type);
}

/** Used by router when inferring command risk from type alone. */
export function isHighRiskCommandType(type: RemoteCommandType): boolean {
  return HIGH_RISK_COMMAND_TYPES.has(type);
}

File: integrations/old-source-preservation/ivault-eye-tracking/snapshot/lib/features/remote/remote_policy.dart
import 'remote_commands.dart';
import 'remote_types.dart';

/// Gaze cannot execute these (spec §18).
const Set<String> kGazeBlockedCommandTypes = {
  RemoteCommandTypes.withdraw,
  RemoteCommandTypes.pay,
  RemoteCommandTypes.tip,
  RemoteCommandTypes.publishCampaign,
  RemoteCommandTypes.connectPlatform,
  RemoteCommandTypes.disconnectPlatform,
  RemoteCommandTypes.convertCoins,
};

/// Mutable policy snapshot for MVP demos and tests.
final class RemotePolicyContext {
  RemotePolicyContext({
    this.ageGroup = 'adult',
    this.kycVerified = true,
    this.trustTier = 2,
    this.walletLocked = false,
    this.fraudHoldActive = false,
    this.campaignBudgetReady = true,
  });

  /// `'minor'` or `'adult'`.
  String ageGroup;
  bool kycVerified;
  int trustTier;
  bool walletLocked;
  bool fraudHoldActive;
  bool campaignBudgetReady;
}

/// Result of [evaluateRemotePermission].
final class RemotePermissionResult {
  const RemotePermissionResult({
    required this.allowed,
    this.reason,
    this.requiresConfirmation = false,
  });

  final bool allowed;
  final String? reason;
  final bool requiresConfirmation;
}

RemotePermissionResult evaluateRemotePermission(
  RemoteCommand command,
  RemotePolicyContext policy, {
  required bool remoteLocked,
}) {
  if (remoteLocked && command.type != RemoteCommandTypes.unlockRemote) {
    return const RemotePermissionResult(
      allowed: false,
      reason: 'Remote is locked.',
    );
  }

  if (policy.ageGroup == 'minor' && command.requiresAdult) {
    return const RemotePermissionResult(
      allowed: false,
      reason: 'This action is not available for this account.',
    );
  }

  if (command.requiresKyc && !policy.kycVerified) {
    return const RemotePermissionResult(
      allowed: false,
      reason: 'Identity verification required.',
    );
  }

  if (command.trustTierRequired != null &&
      policy.trustTier < command.trustTierRequired!) {
    return const RemotePermissionResult(
      allowed: false,
      reason: 'Trust tier too low.',
    );
  }

  if (policy.fraudHoldActive) {
    return const RemotePermissionResult(
      allowed: false,
      reason: 'Action paused by safety review.',
    );
  }

  if (command.type == RemoteCommandTypes.publishCampaign &&
      !policy.campaignBudgetReady) {
    return const RemotePermissionResult(
      allowed: false,
      reason: 'Campaign budget not ready.',
    );
  }

  if ((command.type == RemoteCommandTypes.pay ||
          command.type == RemoteCommandTypes.withdraw ||
          command.type == RemoteCommandTypes.tip) &&
      policy.walletLocked) {
    return const RemotePermissionResult(
      allowed: false,
      reason: 'Wallet is locked.',
    );
  }

  if (command.inputSource == RemoteInputSource.gaze &&
      kGazeBlockedCommandTypes.contains(command.type)) {
    return const RemotePermissionResult(
      allowed: false,
      reason: 'This action cannot run from gaze input.',
    );
  }

  final spec = kRemoteCommandRegistry[command.type];
  final specRisk = spec?.riskLevel ?? command.riskLevel;
  final requiresConfirmation = command.riskLevel == RemoteRiskLevel.high ||
      command.riskLevel == RemoteRiskLevel.medium ||
      specRisk == RemoteRiskLevel.high ||
      specRisk == RemoteRiskLevel.medium ||
      command.requiresConfirmation ||
      (spec?.requiresConfirmation ?? false);

  return RemotePermissionResult(
    allowed: true,
    requiresConfirmation: requiresConfirmation,
  );
}
